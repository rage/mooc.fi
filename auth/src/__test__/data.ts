import { NextFunction, Request, Response } from "express"

export const testProfile = {
  edu_person_affiliation: "",
  edu_person_principal_name: "test",
  schac_home_organization: "helsinki.fi",
  organizational_unit: "organizational_unit",
  schac_personal_unique_code:
    "urn:schac:personalUniqueCode:int:studentID:helsinki.fi:121345678",
  display_name: "test",
  given_name: "test",
  mail: "test@mail.com",
}

export const req = {
  body: {},
  query: {
    RelayState: JSON.stringify({
      language: "fi",
      provider: "hy",
      action: "sign-in",
    }),
  },
  params: {},
  login: jest.fn().mockImplementation((_: any, fn: Function) => fn(null)),
  logout: jest.fn(),
} as unknown as Request
export const res = {
  redirect: jest.fn().mockReturnThis(),
  locals: {},
  setMOOCCookies: jest.fn().mockReturnThis(),
} as unknown as Response
export const next = jest.fn() as NextFunction

export const okTokenResponse = {
  data: {
    status: 200,
    success: true,
    tmc_token: "tmc_token",
    access_token: "access_token",
    admin: false,
  },
}
