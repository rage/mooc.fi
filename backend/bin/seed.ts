/*import {
  Prisma,
  CourseCreateInput,
  CourseStatus,
  CourseTranslationCreateWithoutCourseInput,
  CourseTranslationCreateManyWithoutCourseInput,
  StudyModuleCreateInput,
  StudyModuleTranslationCreateWithoutStudy_moduleInput,
  StudyModuleTranslationCreateManyWithoutStudy_moduleInput,
  StudyModuleWhereUniqueInput,
  CourseTranslation,
} from "../generated/prisma-client"*/
import { PrismaClient, course_status, course_translation } from "@prisma/client"

const Modules = [
  {
    id: "21d838c9-54c7-4fbb-8e4c-91e8f14f0d73",
    slug: "webdev",
    name: "Web development",
    image: "WebModule.jpg",
    study_module_translation: [
      {
        id: "0616fba6-e119-4d77-9b71-2b0fae2101b3",
        language: "en_US",
        name: "Web development",
        description: "Become a super professional web developer and unicorn",
      },
      {
        id: "948698ce-837c-4e0e-93f6-679ea2688691",
        language: "fi_FI",
        name: "Verkko-ohjelmointi",
        description: "Opi ohjelmoimaan verkkosovelluksia ja ole yksisarvinen",
      },
    ],
  },
  {
    id: "662f537e-4395-40db-a32d-710b51fa169e",
    slug: "program",
    name: "Programming skills",
    image: "CodeModule.jpg",
    study_module_translation: [
      {
        id: "9fa46739-8993-4b32-8624-00d30e47830b",
        language: "fi_FI",
        name: "Ohjelmointitaidot",
        description: "Opi ohjelmoitsemaan taitavasti.",
      },
      {
        id: "b6781675-4b03-4bfa-b717-ae40f4262b74",
        language: "en_US",
        name: "Programming skills",
        description: "Learn programming.",
      },
    ],
  },
]

const Courses = [
  {
    id: "jdoifsbf",
    name: "Ohjelmoinnin MOOC 2019",
    slug: "ohpe",
    course_translation: [
      {
        name: "Ohjelmoinnin MOOC 2019",
        language: "fi_FI",
        description:
          "Ohjelmointia Javalla perusteista lähtien sekä mahdollisuus opinto-oikeuteen. Täydellinen kurssi ohjelmoinnin alkeiden opetteluun. Ei vaadi esitietoja.",
        link: "https://ohjelmointi-19.mooc.fi/",
      },
    ],
    promote: true,
    status: "Active",
    start_point: true,
    study_module: ["662f537e-4395-40db-a32d-710b51fa169e"],
  },
  {
    id: "dgoijdgoijf8904",
    name: "Full stack open 2019",
    slug: "fullstack",
    course_translation: [
      {
        id: "idojgf4903",
        name: "Full stack open 2019",
        language: "fi_FI",
        description:
          "Syväsukellus moderniin websovelluskehitykseen — ota haltuusi React, Redux, Node.js, GraphQL ja MongoDB! Kurssi alkaa 15.3. ja päättyy 15.12.",
        link: "https://fullstackopen.com/",
      },
      {
        id: "djgisopgjiotrx",
        name: "Full stack open 2019",
        language: "en_US",
        description:
          "Deep dive into modern web development — learn to master React, Redux, Node.js, GraphQL and MongoDB!",
        link: "https://fullstackopen.com/en/",
      },
    ],
    promote: false,
    status: "Active",
    start_point: true,
    study_module: ["21d838c9-54c7-4fbb-8e4c-91e8f14f0d73"],
  },
  {
    id: "tfujhgfdgrs",
    name: "Elements of AI",
    slug: "elements-of-ai",
    course_translation: [
      {
        id: "ifghdfdh",
        name: "Elements of AI",
        language: "fi_FI",
        description:
          "Vievätkö robotit työmme? Miten tekoäly muuttaa työelämää seuraavan 10 vuoden aikana? Millainen yhteiskunnallinen vaikutus tekoälyllä on?",
        link: "https://www.elementsofai.com/fi",
      },
      {
        id: "fsysrtyrzsj",
        name: "Elements of AI",
        language: "en_US",
        description:
          "Will robots take our jobs? How will artificial intelligence change our working life in the next 10 years? How will intelligence impact our society?",
        link: "https://www.elementsofai.com/",
      },
    ],
    promote: true,
    status: "Active",
    start_point: true,
  },
  {
    id: "hftsgtraet",
    name: "Data Analysis with Python",
    slug: "data-analysis-python",
    course_translation: [
      {
        id: "gdstreedst",
        name: "Data Analysis with Python",
        language: "fi_FI",
        description:
          "Course gives a practical introduction to data analysis using a large number of programming exercises and a project delving into the realm of a selected field of science. Starts on June 19th.",
        link:
          "https://saskeli.github.io/data-analysis-with-python-summer-2019/",
      },
      {
        id: "sefdtgw4ya",
        name: "Data Analysis with Python",
        language: "en_US",
        description:
          "Course gives a practical introduction to data analysis using a large number of programming exercises and a project delving into the realm of a selected field of science. Starts on June 19th.",
        link:
          "https://saskeli.github.io/data-analysis-with-python-summer-2019/",
      },
    ],
    promote: false,
    status: "Active",
    start_point: true,
  },
  {
    id: "hfhdrr",
    name: "Web-palvelinohjelmointi Java",
    slug: "wepa",
    course_translation: [
      {
        id: "jyr6876",
        name: "Web-palvelinohjelmointi JAVA",
        language: "fi_FI",
        description:
          "Java-kielisten web-sovellusten toteutus tutuksi. Opi mm. Spring sovelluskehyksen periaatteet.Kurssi on käynnissä. Ensimmäisen osan tehtävien deadline on 18.3. Kurssi päättyy 10.5.",
        link: "https://web-palvelinohjelmointi-19.mooc.fi/",
      },
    ],
    promote: false,
    status: "Ended",
    start_point: false,
    study_module: ["21d838c9-54c7-4fbb-8e4c-91e8f14f0d73"],
  },
  {
    id: "gsrhrthrs",
    name: "Johdatus tietoliikenteeseen",
    slug: "johto",
    course_translation: [
      {
        id: "fdnjeso",
        name: "Johdatus tietoliikenteeseen",
        language: "fi_FI",
        description:
          "Miten internet toimii? Opi miten tietokoneet, puhelimet ja palvelimet keskustelevat verkon yli. Mitä ovat reitittimet ja kytkimet?",
        link: "https://johdatus-tietoliikenteeseen-19.mooc.fi/",
      },
    ],
    promote: true,
    status: "Ended",
    start_point: true,
  },
  {
    id: "es646rki",
    name: "DevOps with Docker",
    slug: "docker",
    course_translation: [
      {
        id: "545thjftfhes",
        name: "DevOps with Docker",
        language: "fi_FI",
        description:
          "Miten sovellusten käyttöönotto tapahtuu nykyään? Tällä kurssilla tutustutaan Dockeriin ja docker-composeen. Samalla opit monista erilaisista osista joista webpalvelut koostuvat.",
        link: "https://docker-hy.github.io/",
      },
      {
        id: "ewiau8a43",
        name: "DevOps with Docker",
        language: "en_US",
        description:
          "Learn the basics of modern software deployment in this introductory course to Docker and docker-compose. Explore different parts of web services, such as reverse proxies, databases etc.",
        link: "https://docker-hy.github.io/",
      },
    ],
    promote: false,
    status: "Active",
    start_point: true,
  },
  {
    id: "egdfhftuer65745",
    name: "Tietokantojen perusteet",
    slug: "tikape",
    course_translation: [
      {
        id: "gdjiaoter",
        name: "Tietokantojen perusteet",
        language: "fi_FI",
        description:
          "Miten tietokannat toimivat? Opi SQL-kielen perusteet, relaatietokantojen suunnittelun ja käytön, sekä tietokantaa käyttävien ohjelmistojen perusteet.",
        link: "https://tietokantojen-perusteet-19.mooc.fi/",
      },
    ],
    promote: false,
    status: "Ended",
    start_point: true,
    study_module: ["662f537e-4395-40db-a32d-710b51fa169e"],
  },
  {
    id: "dhyrrtutrutrik",
    name: "Tietorakenteet ja algoritmit",
    slug: "tira",
    course_translation: [
      {
        id: "gdfgtewt",
        name: "Tietorakenteet ja algoritmit",
        language: "fi_FI",
        description:
          "Osaat jo ohjelmoida mutta haluat kehittyä lisää. Opi suunnittelemaan tehokkaita algoritmeja, joilla voit ratkaista kinkkisiä ongelmia salamannopeasti. Kurssi alkaa näillä näkymin syksyllä 2019.",
        link: "https://tietokantojen-perusteet-19.mooc.fi/",
      },
    ],
    promote: false,
    status: "Upcoming",
    start_point: true,
    study_module: ["662f537e-4395-40db-a32d-710b51fa169e"],
  },
  {
    id: "290e6506-00e3-433d-b938-1a258b6b04f2",
    name: "Tietokoneen toiminnan perusteet",
    slug: "tito",
    course_translation: [
      {
        id: "fsdedhbdf",
        name: "Tietokoneen toiminnan perusteet",
        language: "fi_FI",
        description:
          "Mitä tietokoneet ovat ja miten ne toimivat? Opi alkeet millaisista osista ja toiminnallisuuksista nämä meille elintärkeät laitteet koostuvat.",
        link: "",
      },
    ],
    promote: false,
    status: "Upcoming",
    start_point: true,
    study_module: ["662f537e-4395-40db-a32d-710b51fa169e"],
  },
  {
    id: "hfdhgdrdeetteet4et4",
    name: "Cyber Security Base",
    slug: "cyber",
    course_translation: [
      {
        id: "fsjieojewoirtjoi",
        name: "Cyber Security Base",
        language: "fi_FI",
        description:
          "Become a cyber security professional. Cyber Security Base focuses on building core knowledge and abilities related to the work of a cyber security professional. Starts on October 28th.",
        link: "",
      },
      {
        id: "gndjrozipa5390",
        name: "Cyber Security Base",
        language: "en_US",
        description:
          "Become a cyber security professional. Cyber Security Base focuses on building core knowledge and abilities related to the work of a cyber security professional. Starts on October 28th.",
        link: "",
      },
    ],
    promote: false,
    status: "Upcoming",
    start_point: true,
  },
  {
    id: "sgjdeirotueori",
    name: "Ohjelmoinnin MOOC 2018",
    slug: "ohpe18",
    course_translation: [
      {
        id: "jfsjoidfoijdfoijdfsoijdfsoij",
        name: "Ohjelmoinnin MOOC 2018",
        language: "fi_FI",
        description:
          "Ohjelmointia Javalla perusteista lähtien sekä mahdollisuus opinto-oikeuteen. Täydellinen kurssi ohjelmoinnin alkeiden opetteluun. Ei vaadi esitietoja.",
        link: "",
      },
    ],
    promote: false,
    status: "Ended",
    start_point: true,
    study_module: ["662f537e-4395-40db-a32d-710b51fa169e"],
  },
  {
    id: "4545thythyggh",
    name: "Cyber Security Base(2018)",
    slug: "cyber2018",
    course_translation: [
      {
        id: "4243trhftghf",
        name: "Cyber Security Base(2018)",
        language: "fi_FI",
        description:
          "Become a cyber security professional. Cyber Security Base focuses on building core knowledge and abilities related to the work of a cyber security professional.",
        link: "https://cybersecuritybase.mooc.fi/",
      },
      {
        id: "75tyjfxkhmx,kgxmj",
        name: "Cyber Security Base(2018)",
        language: "en_US",
        description:
          "Become a cyber security professional. Cyber Security Base focuses on building core knowledge and abilities related to the work of a cyber security professional.",
        link: "https://cybersecuritybase.mooc.fi/",
      },
    ],
    promote: false,
    status: "Ended",
    start_point: true,
  },
  {
    id: "<rwjogh<ugrixtjö",
    name: "2013 Programming with Java I",
    slug: "ohpe13",
    course_translation: [
      {
        id: "75tyjfxkhmx,kgxmj",
        name: "2013 Programming with Java I",
        language: "en_US",
        description:
          "Learn the basics of computer programming, algorithms and object-oriented programming using the Java programming language.",
        link: "http://moocfi.github.io/courses/2013/programming-part-1/",
      },
    ],
    promote: true,
    status: "Active",
    start_point: true,
    study_module: ["662f537e-4395-40db-a32d-710b51fa169e"],
  },
  {
    id: "<rwjghrsyt",
    name: "2013 Programming with Java II",
    slug: "ohpe13-2",
    course_translation: [
      {
        id: "75tyjfxkhmx,kgxmj",
        name: "2013 Programming with Java II",
        language: "en_US",
        description:
          "This course is a direct continuation of the course Object-Oriented Programming with Java, part I.",
        link: "http://moocfi.github.io/courses/2013/programming-part-2/",
      },
    ],
    promote: false,
    status: "Active",
    start_point: true,
    study_module: ["662f537e-4395-40db-a32d-710b51fa169e"],
  },
]

const prisma = new PrismaClient()

const seed = async () => {
  await prisma.study_module.deleteMany({ where: {} })
  await prisma.course.deleteMany({ where: {} })

  await Promise.all(
    Modules.map(async (module) => {
      const _module = {
        ...module,
        study_module_translation: module.study_module_translation
          ? {
              create: (module.study_module_translation || []).map((t) => ({
                ...t,
                id: undefined,
              })),
            }
          : null,
      }

      return await prisma.study_module.create({
        data: _module,
      })
    }),
  )

  return await Promise.all(
    Courses.map(async (course) => {
      const _course = {
        ...course,
        id: undefined,
        photo: undefined,
        teacher_in_charge_name: "",
        teacher_in_charge_email: "",
        start_date: "",
        status: course.status as course_status,
        course_translation: course.course_translation
          ? {
              create:
                (course?.course_translation as course_translation[])?.map(
                  (t) => ({
                    ...t,
                    id: undefined,
                    link: t.link || "",
                  }),
                ) ?? undefined,
            }
          : null,
        study_module: null,
      }

      const newCourse = await prisma.course.create({ data: _course })
      if (course.study_module) {
        await prisma.course.update({
          where: {
            id: newCourse.id,
          },
          data: {
            study_module: {
              connect: course.study_module.map((id) => ({
                id,
              })),
            },
          },
        })
      }

      return Promise.resolve()
    }),
  )
}

seed()
