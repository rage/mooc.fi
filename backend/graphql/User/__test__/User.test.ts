import { gql } from "graphql-request"
import { getTestContext, fakeTMCCurrent } from "../../../tests/__helpers"
import {
  adminUserDetails,
  normalUser,
  normalUserDetails,
} from "../../../tests/data"
import { seed } from "../../../tests/data/seed"
import { orderBy } from "lodash"

const addUserMutation = gql`
  mutation AddUser($user: UserArg!) {
    addUser(user: $user) {
      id
      email
      first_name
      last_name
      username
      research_consent
      upstream_id
      administrator
    }
  }
`

const updateReseachConsentMutation = gql`
  mutation UpdateResearchConsent($value: Boolean!) {
    updateResearchConsent(value: $value) {
      id
    }
  }
`

const updateUserNameMutation = gql`
  mutation updateUserName($first_name: String, $last_name: String) {
    updateUserName(first_name: $first_name, last_name: $last_name) {
      id
      first_name
      last_name
    }
  }
`

const ctx = getTestContext()
const tmc = fakeTMCCurrent({
  "Bearer normal": [200, normalUserDetails],
  "Bearer admin": [200, adminUserDetails],
})

// @ts-ignore: not used
const sortByExercise = (data: any) => {
  if (!data?.exercise_completions) {
    return data
  }

  return {
    ...data,
    exercise_completions: orderBy(
      data.exercise_completions.map((ec: any) => ec.exercise),
      ["id"],
      ["asc"],
    ),
  }
}

describe("User", () => {
  describe("queries", () => {
    beforeEach(() => {
      tmc.setup()
    })

    afterAll(() => tmc.teardown())

    describe("currentUser", () => {
      beforeEach(async () => {
        await seed(ctx.prisma)
      })
      /*beforeEach(async () => {
        await ctx.prisma.user.create({
          data: normalUser,
        })
      })*/

      it("shows current user when logged in", async () => {
        ctx!.client.setHeader("Authorization", "Bearer normal")

        const res = await ctx.client.request(`
      query {
        currentUser {
          id
          administrator
          email
          first_name
          last_name
          username
          upstream_id
        }
      }
    `)

        expect(res).toMatchInlineSnapshot(
          {
            currentUser: {
              id: expect.any(String),
              administrator: false,
              email: "e@mail.com",
              first_name: "first",
              last_name: "last",
              username: "user",
              upstream_id: 1,
            },
          },
          `
              Object {
                "currentUser": Object {
                  "administrator": false,
                  "email": "e@mail.com",
                  "first_name": "first",
                  "id": Any<String>,
                  "last_name": "last",
                  "upstream_id": 1,
                  "username": "user",
                },
              }
          `,
        )

        jest.clearAllMocks()
      })

      it("shows null when not logged in", async () => {
        ctx!.client.setHeader("Authorization", "")

        const res = await ctx.client.request(`
      query {
        currentUser {
          id
          administrator
          email
          first_name
          last_name
          username
          upstream_id
        }
      }
    `)

        expect(res).toMatchInlineSnapshot(
          {
            currentUser: null,
          },
          `
        Object {
          "currentUser": null,
        }
      `,
        )
      })
    })

    describe("user_course_summary", () => {
      beforeEach(async () => {
        await seed(ctx.prisma)
      })

      it("returns courses with completions", async () => {
        ctx!.client.setHeader("Authorization", "Bearer normal")

        const res = await ctx.client.request(`
          query {
            currentUser {
              id
              user_course_summary {
                user_id
                course_id
                course {
                  id
                  name
                  exercises {
                    id
                    name
                    course_id
                    part
                    section
                    max_points
                  }
                }
                exercise_completions {
                  id
                  exercise_id
                  timestamp
                  n_points
                  attempted
                  completed
                  exercise_completion_required_actions {
                    id
                    value
                  }
                }
              }
            }
          }
        `)

        const sortedRes = {
          currentUser: {
            ...res?.currentUser,
            user_course_summary: [
              ...res?.currentUser?.user_course_summary?.map((cs: any) => ({
                ...cs,
                course: {
                  ...cs.course,
                  exercises: orderBy(cs.course.exercises, ["name"]),
                },
                exercise_completions: orderBy(cs?.exercise_completions, ["id"]),
              })),
            ],
          },
        }

        expect(sortedRes).toMatchSnapshot()
      })
    })
  })

  describe("mutations", () => {
    beforeAll(() => {
      tmc.setup()
    })

    afterAll(() => tmc.teardown())

    beforeEach(async () => {
      await ctx.prisma.user.deleteMany({ where: {} })
    })

    describe("addUser", () => {
      beforeAll(() => ctx!.client.setHeader("Authorization", ""))

      it("creates user correctly", async () => {
        const res = await ctx.client.request(addUserMutation, {
          user: {
            email: "e@mail.com",
            first_name: "first",
            last_name: "last",
            username: "username",
            research_consent: false,
            upstream_id: 1,
          },
        })

        expect(res).toMatchSnapshot({
          addUser: {
            id: expect.any(String),
          },
        })
      })

      it("won't create user with same id", async () => {
        await ctx.prisma.user.create({
          data: normalUser,
        })

        await expect(async () => {
          await ctx.client.request(addUserMutation, {
            user: {
              email: "e@mail.com",
              first_name: "first",
              last_name: "last",
              username: "username",
              research_consent: false,
              upstream_id: 1,
            },
          })
        }).rejects.toThrow()
      })
    })

    describe("updateResearchConsent", () => {
      beforeEach(async () => {
        await ctx!.prisma.user.create({
          data: normalUser,
        })
      })

      afterEach(async () => {
        await ctx!.prisma.user.delete({ where: { upstream_id: 1 } })
        ctx.user = undefined
      })

      it("updates correctly", async () => {
        ctx!.client.setHeader("Authorization", "Bearer normal")

        const res = await ctx!.client.request(updateReseachConsentMutation, {
          value: true,
        })

        expect(res.updateResearchConsent).toMatchInlineSnapshot(
          { id: expect.any(String) },
          `
        Object {
          "id": Any<String>,
        }
      `,
        )
        const updatedConsent = await ctx!.prisma.user.findFirst({
          where: { upstream_id: 1 },
          select: { research_consent: true },
        })

        expect(updatedConsent).toMatchObject({ research_consent: true })
      })

      it("won't update research consent without auth", async () => {
        ctx!.client.setHeader("Authorization", "")

        try {
          await ctx!.client.request(updateReseachConsentMutation, {
            value: true,
          })
          fail()
        } catch {}
      })
    })

    describe("updateUserName", () => {
      beforeEach(async () => {
        await ctx!.prisma.user.create({
          data: normalUser,
        })
      })

      it("updates correctly", async () => {
        ctx!.client.setHeader("Authorization", "Bearer normal")

        const res = await ctx!.client.request(updateUserNameMutation, {
          first_name: "updated first",
          last_name: "updated last",
        })

        expect(res.updateUserName).toMatchObject({
          first_name: "updated first",
          last_name: "updated last",
        })
      })

      it("errors without auth", async () => {
        ctx!.client.setHeader("Authorization", "")
        try {
          await ctx!.client.request(updateUserNameMutation, {})
          fail()
        } catch {}
      })
    })
  })
})
