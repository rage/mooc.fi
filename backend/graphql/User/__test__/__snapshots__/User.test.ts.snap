// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User mutations addUser creates user correctly 1`] = `
Object {
  "addUser": Object {
    "administrator": false,
    "email": "e@mail.com",
    "first_name": "first",
    "id": Any<String>,
    "last_name": "last",
    "research_consent": false,
    "upstream_id": 1,
    "username": "username",
  },
}
`;

exports[`User queries user_course_summary returns courses with completions 1`] = `
Object {
  "currentUser": Object {
    "id": "20000000-0000-0000-0000-000000000102",
    "user_course_summary": Array [
      Object {
        "course": Object {
          "exercises": Array [
            Object {
              "course_id": "00000000-0000-0000-0000-000000000002",
              "id": "50000000-0000-0000-0000-000000000003",
              "max_points": 3,
              "name": "exercise 3",
              "part": null,
              "section": null,
            },
          ],
          "id": "00000000-0000-0000-0000-000000000002",
          "name": "course1",
        },
        "course_id": "00000000-0000-0000-0000-000000000002",
        "exercise_completions": Array [
          Object {
            "attempted": true,
            "completed": false,
            "exercise_completion_required_actions": Array [
              Object {
                "id": "66666666-0000-0000-0000-000000000003",
                "value": "TOO_MANY_DUCKS",
              },
            ],
            "exercise_id": "50000000-0000-0000-0000-000000000003",
            "id": "60000000-0000-0000-0000-000000000003",
            "n_points": 4,
            "timestamp": "1900-01-01T08:00:00.000Z",
          },
          Object {
            "attempted": true,
            "completed": false,
            "exercise_completion_required_actions": Array [
              Object {
                "id": "66666666-0000-0000-0000-000000000004",
                "value": "existing",
              },
            ],
            "exercise_id": "50000000-0000-0000-0000-000000000004",
            "id": "60000000-0000-0000-0000-000000000004",
            "n_points": 1,
            "timestamp": "2000-01-01T08:00:00.000Z",
          },
        ],
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
      Object {
        "course": Object {
          "exercises": Array [
            Object {
              "course_id": "00000000-0000-0000-0000-000000000001",
              "id": "50000000-0000-0000-0000-000000000001",
              "max_points": 2,
              "name": "exercise 1",
              "part": 1,
              "section": 1,
            },
            Object {
              "course_id": "00000000-0000-0000-0000-000000000001",
              "id": "50000000-0000-0000-0000-000000000002",
              "max_points": 3,
              "name": "exercise 2",
              "part": 1,
              "section": 2,
            },
          ],
          "id": "00000000-0000-0000-0000-000000000001",
          "name": "course2",
        },
        "course_id": "00000000-0000-0000-0000-000000000001",
        "exercise_completions": Array [
          Object {
            "attempted": null,
            "completed": true,
            "exercise_completion_required_actions": Array [],
            "exercise_id": "50000000-0000-0000-0000-000000000001",
            "id": "60000000-0000-0000-0000-000000000001",
            "n_points": 6,
            "timestamp": "1900-01-01T08:00:00.000Z",
          },
          Object {
            "attempted": null,
            "completed": false,
            "exercise_completion_required_actions": Array [],
            "exercise_id": "50000000-0000-0000-0000-000000000002",
            "id": "60000000-0000-0000-0000-000000000002",
            "n_points": 0,
            "timestamp": "1900-01-01T08:00:00.000Z",
          },
        ],
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
    ],
  },
}
`;
